File: 0121-best-time-to-buy-and-sell-stock/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int maxProfit = 0;
        int cost = prices[0];//7

        for(int i=1; i<n; i++) {
            // cheapest price till today
            if(prices[i] < cost) {
                // buy -> update the cheapest buy
                cost = prices[i];//1
            } else {
                // sell today
                maxProfit = Math.max(maxProfit, prices[i] - cost);
            }
        }
        return maxProfit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/solution.swift
class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
        var profit = 0
        var minVal = prices.max()!
        var n = prices.count

        for i in 0..<n {
            if prices[i] < minVal {
                minVal = prices[i]
            } else {
                profit = max(profit, prices[i] - minVal)//4
            }
        }
        return profit
    }
}
